[HigherOrderOp] add pytree operands tests for cond

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112661

This is a follow-up of #111611. After this PR, we allow pytree with tensor-only leaves as operands of branches.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @zou3519

ydwu4 committed:
[HigherOrderOp] add pytree operands tests for cond

[ghstack-poisoned]

ydwu4 committed:
Update on "[HigherOrderOp] add pytree operands tests for cond"


This is a follow-up of #111611. Right now, we only allow pytree with tensor-only leaves as operands of branches.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng zou3519

[ghstack-poisoned]

zou3519 commented on test/dynamo/test_export.py:
```diff
@@ -3023,7 +3023,7 @@ def f_non_list_operands(x):
         with self.assertRaisesRegex(
             RuntimeError,
-            "Expect operands to be a tuple of Tensors, but got",
+            r"Expect operands to be a tuple of pytrees that only consists of tensor leaves, but got",
```
nit: The word "pytree" may confuse users. A better error message is something like "Expected operands to only contain Tensors"

zou3519 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -75,7 +75,7 @@ def cond(pred, true_branch, false_branch, operands):
           the same, and the outputs have to be the same type and shape.
 
-        operands (Tuple[torch.Tensor]): A tuple of inputs to the true/false functions.
+        operands (Tuple[Pytree[torch.Tensor]]): A tuple of inputs to the true/false functions.
```
Ditto on the pytree feedback here: "pytree" is our terminology for "possibly nested collection of tensors", not sure if users will know what's up

ydwu4 replied:
yeah, let me think about how to use a different approach of describing it.

zou3519 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -130,10 +128,10 @@ def _validate_input(pred, true_fn, false_fn, operands):
 
         if not isinstance(operands, (tuple, list)) or any(
-            not isinstance(t, torch.Tensor) for t in operands
+            not isinstance(t, torch.Tensor) for t in pytree.tree_flatten(operands)[0]
```
pytree.tree_flatten(operands)[0] -> pytree.tree_leaves(operands)

ydwu4 replied:
yes, i can change to tree_any!

> Also is the not isinstance(operands, (tuple, list)) part of the condition still valid?

Yeah, that's one point I want to discuss:
- JAX's cond's signature is `cond(pred, true_fn, false_fn, *operands)`, where operands are positional-only arguments. 
- What we have now is `cond(pred, true_fn, false_fn, operands)`, where operands is a tuple of tensors. Our approach is a little more vobose because users need to wrap inputs with `()` but it looks more structured. 
- The TF approach is `cond(pred, true_fn, false_fn)`, where true_fn and false_fn takes no argument and solely relies on closure.

I have no strong opinion on which one to use. If we'd like to change the signature to be consistent with JAX's or TF's, I can do it in the follow-up pr just that there are quite a number of tests we need to change (i.e. unrolling the input tuple).

zou3519 replied:
We should just do ours. We went through an exercise a while ago and determined that we should do (operands) to be consistent with how other control flow op APIs should look like, right?

ydwu4 replied:
yeah! i recall our discussion.

zou3519 reviewed (approved):
Some minor comments, otherwise LGTM

ydwu4 committed:
Update on "[HigherOrderOp] add pytree operands tests for cond"


This is a follow-up of #111611. After this PR, we allow pytree with tensor-only leaves as operands of branches.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng zou3519

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 committed:
Update on "[HigherOrderOp] add pytree operands tests for cond"


This is a follow-up of #111611. After this PR, we allow pytree with tensor-only leaves as operands of branches.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng zou3519

[ghstack-poisoned]
